### PR TITLE
AAP-56029 Collect test coverage from everything in the matrix

### PR DESCRIPTION
**Pull request by AlanCoding**
## Description
Collect coverage from all jobs in the tox / test matrix.

## Type of Change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [x] Test update
- [ ] Refactoring (no functional changes)
- [ ] Development environment change
- [ ] Configuration change

## Self-Review Checklist
- [ ] I have performed a self-review of my code
- [ ] I have added relevant comments to complex code sections
- [ ] I have updated documentation where needed
- [ ] I have considered the security impact of these changes
- [ ] I have considered performance implications
- [ ] I have thought about error handling and edge cases
- [ ] I have tested the changes in my local environment

## Testing Instructions
This, itself, modifies testing. Not even testing, collection of coverage of tests. Not even changing whether coverage is collected, but how thoroughly coverage is collected, making it more thorough.

### Prerequisites
none

### Steps to Test
This is _about_ testing.

### Expected Results
coverage for entire matrix collected

## Additional Context
Created because it came up in https://github.com/ansible/django-ansible-base/pull/855, and separated out because that is a highly-blocking issue that we should not scope creep.

### Required Actions
- [ ] Requires documentation updates
- [ ] Requires downstream repository changes
- [ ] Requires infrastructure/deployment changes
- [ ] Requires coordination with other teams
- [ ] Blocked by PR/MR: #XXX

### Screenshots/Logs
See results of CI for this PR

---

> [!NOTE]
> Collects coverage from all tox matrix jobs, uploads per-env artifacts, aggregates them in Sonar workflows, and passes coverage paths to SonarCloud for PR and branch analyses.
> 
> - **CI (`.github/workflows/ci.yml`)**:
>   - Inject PR number into `coverage.xml` only when `matrix.tests.env != 'check'` and file exists.
>   - Upload coverage artifact per env as `coverage-${{ matrix.tests.env }}` (instead of single `coverage`).
> - **SonarCloud (`.github/workflows/sonar-pr.yml`)**:
>   - Download all coverage artifacts with pattern `coverage-*` for both PR and branch analyses.
>   - Discover all `coverage.xml` files, set `COVERAGE_PATHS`, and validate presence for PR analysis.
>   - Extract PR number from the first discovered `coverage.xml`.
>   - Pass `-Dsonar.python.coverage.reportPaths=${{ env.COVERAGE_PATHS }}` to SonarCloud (conditionally for branch analysis).
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit f7baba0200d5d89143144266bc7849ccc624c5dd. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: .github/workflows/sonar-pr.yml [hunk 1]
```diff
@@ -61,14 +61,20 @@ jobs:
           HEAD_BRANCH: ${{ github.event.workflow_run.head_branch }}
           GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
         run: |
+          # Find all downloaded coverage XML files
+          coverage_files=$(find . -name "coverage-*.xml" -type f | tr '\n' ',' | sed 's/,$//')
+          echo "Found coverage files: $coverage_files"
```
[the remark this answers is not in the document]
Comment (AlanCoding) [reply .github/workflows/sonar-pr.yml:66]: ok this seems to be the real issue

**Comment by AlanCoding** [comment]
Current outputs:

Artifact coverage-py310.zip successfully finalized. Artifact ID 4278103454
Artifact coverage-py310 has been successfully uploaded! Final size is 37982 bytes. Artifact ID is 4278103454
Artifact download URL: https://github.com/ansible/django-ansible-base/actions/runs/18532807655/artifacts/4278103454

and indeed we have the artifacts

[coverage-py310](https://github.com/ansible/django-ansible-base/actions/runs/18532807655/artifacts/4278103454)
37.1 KB	
sha256:8d5fe9192dfe2535848d3d80ada8454fc25c0b762fc602a343bff811789cfca9
[coverage-py311](https://github.com/ansible/django-ansible-base/actions/runs/18532807655/artifacts/4278087333)
37.1 KB	
sha256:c3cbd75bc8faa4daa0bb541c7948dbac24ed1bd74ba878ea9f02990c2a932577
[coverage-py311sqlite](https://github.com/ansible/django-ansible-base/actions/runs/18532807655/artifacts/4278097982)
37.3 KB	
sha256:5ae6269f64d30339956277480b17445ed8747c9051819ca26fb7adc634323f83
[coverage-py312](https://github.com/ansible/django-ansible-base/actions/runs/18532807655/artifacts/4278104815)
37.1 KB	
sha256:7dd089d244887041db46a3e7909a442fe25f260fdf86a42d4958db65563771ac

What went wrong? I don't see any stage where the sonar stuff ran.

**Comment by AlanCoding** [comment]
SonarCloud not reporting and I don't know why

**Review by john-westcott-iv — APPROVED** [review]
This looks correct other than Claudes comment which we should be able to resolve by grabbing the last line of the diff. 

> SonarCloud not reporting and I don't know why

This is one of the "joys" of GH CI. The CI jobs run in the context of the repo the PR is coming from (so your latest code) but the Sonar job runs in the context of the `ansible/django-ansible-base` repo. I'm not sure why Sonar didn't report as failed but it did try to run and you can see the results here: https://github.com/ansible/django-ansible-base/actions/runs/18533132698/job/52821033978.

TL; DR; until this merges it will be out of sync and Sonar will fail. This happens commonly when we work on CI and you basically have to either duplicate the repo into a test repo and then merge your Sonar job into the devel branch directly and then do a pull against the test repo to test. Or feel "pretty sure" that everything is in working order, force merge and then wait to see what happens in the next PR.
